libs/utils/energy_model.py

```diff
@@ -0,0 +1,572 @@
+                           "cpu-sleep","WFI"]
+
+        :param cpus_active: list where bool(cpus_active[N]) is False iff no
+                            tasks will run on CPU N.
+        :returns: List ``ret`` where ``ret[N]`` is the name of the estimated
+                  idle state that CPU N can enter during idle periods.
+
+        """
+        states = self._guess_idle_states(cpus_active)
+        return [s or c.idle_states.keys()[0]
+                for s, c in zip(states, self.cpu_nodes)]
+
+    def _guess_freqs(self, util_distrib):
+        overutilized = False
+        # Find what frequency each CPU would need if it was alone in its
+        # frequency domain
+        ideal_freqs = [0 for _ in self.cpus]
+        for node in self.cpu_nodes:
+            [cpu] = node.cpus
+            required_cap = util_distrib[cpu]
+
+            possible_freqs = [f for f, s in node.active_states.iteritems()
+                              if s.capacity >= required_cap]
+
+            if possible_freqs:
+                ideal_freqs[cpu] = min(possible_freqs)
+            else:
+                # CPU cannot provide required capacity, use max freq
+                ideal_freqs[cpu] = max(node.active_states.keys())
+                overutilized = True
+
+        # Rectify the frequencies among domains
+        freqs = [0 for _ in ideal_freqs]
+        for domain in self.freq_domains:
+            domain_freq = max(ideal_freqs[c] for c in domain)
+            for cpu in domain:
+                freqs[cpu] = domain_freq
+
+        return freqs, overutilized
+
+    def guess_freqs(self, util_distrib):
+        """Work out CPU frequencies required to execute a workload
+
+        Find the lowest possible frequency for each CPU that provides enough
+        capacity to satisfy the utilization, taking into account frequency
+        domains.
+
+        :param util_distrib: Utilization distribution, see
+                             :ref:`util_distribs <util-distrib>`
+        :returns: List ``ret`` where ``ret[N]`` is the frequency that CPU N must
+                  run at
+        """
+        freqs, _ = self._guess_freqs(util_distrib)
+        return freqs
+
+    def _estimate_from_active_time(self, cpu_active_time, freqs, idle_states,
+                                   combine):
+        """Helper for estimate_from_cpu_util
+
+        Like estimate_from_cpu_util but uses active time i.e. proportion of time
+        spent not-idle in the range 0.0 - 1.0.
+
+        If combine=False, return idle and active power as separate components.
+        """
+        power = 0
+        ret = {}
+
+        assert all(0.0 <= a <= 1.0 for a in cpu_active_time)
+
+        for node in self.root.iter_nodes():
+            if not node.active_states or not node.idle_states:
+                continue
+
+            cpus = tuple(node.cpus)
+            # For now we assume topology nodes with energy models do not overlap
+            # with frequency domains
```

Comment:
Sure, I've added a check to `__init__` but I'm also going to leave this comment and assert here, if that sounds good to you.